docs/src/content/docs/guides/clienttoolstutorial.mdx

```diff
@@ -0,0 +1,429 @@
+---
+title: "Tutorial: Building Interactive UI with Client Tools"
+description: Learn how to implement client-side tools in Ultravox to create dynamic, interactive user interfaces.
+---
+import { Steps, Code } from '@astrojs/starlight/components';
+
+This tutorial walks you through implementing client-side tools in Ultravox to create real-time, interactive UI elements. You'll build a drive-thru order display screen that updates dynamically as customers place orders through an AI agent.
+
+**What you'll learn:**
+- How to define and implement client tools
+- Real-time UI updates using custom events
+- State management with React components
+- Integration with Ultravox's AI agent system
+
+**Time to complete:** 30 minutes
+
+## Prerequisites
+
+Before starting this tutorial, make sure you have:
+- Basic knowledge of TypeScript and React
+- The starter code from our [tutorial repository](https://github.com/fixie-ai/ultravox-tutorial-client-tools)
+- Node.js 16+ installed on your machine
+
+## Understanding Client Tools
+
+Client tools in Ultravox enable direct interaction between AI agents and your frontend application. Unlike [server-side tools](../tools) that handle backend operations, client tools are specifically designed for:
+
+- **UI Updates** → Modify interface elements in real-time
+- **State Management** → Handle application state changes
+- **User Interaction** → Respond to and process user actions
+- **Event Handling** → Dispatch and manage custom events
+
+## Project Overview: Dr Donut Drive-Thru
+
+We'll build a drive-thru order display for a fictional restaurant called "Dr. Donut". The display will update in real-time as customers place orders through our AI agent.
+
+### Implementation Steps
+<Steps>
+1. **Define the Tool** → Create a schema for order updates
+1. **Implement Logic** → Build the tool's functionality
+1. **Register the Tool** → Connect it to the Ultravox system
+1. **Create the UI** → Build the order display component
+</Steps>
+
+:::tip[Stuck?]
+If at any point you get lost, you can refer to the `/final` folder in the repo to get final versions of the various files you will create or edit.
+:::
+
+## Step 1: Define the Tool
+
+First, we'll define our `updateOrder` tool that the AI agent will use to modify the order display.
+
+Modify `.demo-config.ts`:
+
+```ts
+const selectedTools: SelectedTool[] = [
+  {
+    "temporaryTool": {
+      "modelToolName": "updateOrder",
+      "description": "Update order details. Used any time items are added or removed or when the order is finalized. Call this any time the user updates their order.",      
+      "dynamicParameters": [
+        {
+          "name": "orderDetailsData",
+          "location": ParameterLocation.BODY,
+          "schema": {
```

Comment:
If this schema is working well for you then great, but I'd expect better model adherence by splitting this into {add,update,remove}OrderedItem so that your nested parameters can be top-level.